deeplearning4j-scaleout/spark/dl4j-spark-ml/src/main/scala/org/deeplearning4j/spark/ml/impl/SparkDl4jNetworkWrapper.scala

```diff
@@ -79,15 +115,28 @@ abstract class SparkDl4jModelWrapper[T, E <: SparkDl4jModelWrapper[T, E]](overri
         } else throw new RuntimeException("Vector size must be greater than 0")
     }
 
-    protected def output(vector: Vector) : Vector = network.predict(vector)
+    protected def output(vector: Vector) : Vector = {
+        val predicted = outputTensor(vector)
+        if (!predicted.isVector) {
+            log.warn("The output of the tensor is not a vector. Flattening the tensor to match Spark's api.")
+        }
+        Vectors.dense(flattenTensor(predicted))
+    }
+
+    def setTrainingStats(sparkTrainingStats: SparkTrainingStats) : this.type = {
+        this.trainingStats = sparkTrainingStats
+        this
+    }
+
+    def getTrainingStats : SparkTrainingStats = trainingStats
 
     protected def outputTensor(vector: Vector) : INDArray = getMultiLayerNetwork.output(MLLibUtil.toVector(vector))
 
     protected def outputFlattenedTensor(vector: Vector) : Vector = Vectors.dense(flattenTensor(outputTensor(vector)))
 
     protected[SparkDl4jModelWrapper] class SparkDl4jModelWriter(instance: SparkDl4jModelWrapper[T,E]) extends MLWriter {
         override protected def saveImpl(path: String): Unit = {
-            ModelSerializer.writeModel(network.getNetwork, path, true)
+            ModelSerializer.writeModel(network, path, true)
         }
```

Comment:
Yes ! This works well, it's actually the approach taken by Spark : to use the HDFS API to access everything, with a `FileSystem` configuration that differs based on the case you are looking at. It often amounts to an import of configuration followed by `val fs = HdfsUtils.getFileSystemForPath(path, hadoopConf)`.

It would indeed be very appropriate to follow suit here.
You can look here for inspiration:
https://github.com/apache/spark/blob/master/streaming/src/main/scala/org/apache/spark/streaming/util/HdfsUtils.scala